### DISCUSSION TITLE
subscribeWithSelector triggers on every store update, even if there is no change

### DISCUSSION THREAD

**Discussion by ivancuric** [pmndrs/zustand#2103] [Bug report] [answered]
Let's say I have the following store:

```ts
const store  = createStore()(subscribeWithSelector(() => {a: 1}))
```

and the following two subscriptions:
```ts
    store.subscribe(
      (x) => x.a,
      (a) => console.log(a),
    );

    store.subscribe((x) => console.log(x));
  ```
  
Updating the store with the same value using `store.setState({a: 1})` will not trigger the first subscription, as expected, but it will trigger the second one.

Is this an expected behaviour?

**Comment by dai-shi** [comment]
If you use `subscribeWithSelector` middleware, the first argument is a selector, not a callback.
Yes, it's an expected behavior.

**Reply by ivancuric** [reply]
Thanks for the fast response!

According to the TS types, this form is still valid:
```ts
store.subscribe((current, previous) => console.log(current))
```
and works the same as this one:
```ts
store.subscribe(
   (store) => store,
   (current, previous) => console.log(current),
 );
```
But let's work with the second option — selecting the entire store.

I'm not sure why selecting the entire store triggers an event when a non-changing update was made, while selecting a property doesn't, that is:
```ts
store.subscribe(
   (store) => store.a,
   (current, previous) => console.log(current),
 );

// doesn't trigger
store.setState({a: store.getState().a})
```

**Reply by dai-shi** [reply]
> According to the TS types, this form is still valid:
> 
> ```ts
> store.subscribe((current, previous) => console.log(current))
> ```

Sorry, that was my bad.

https://github.com/pmndrs/zustand/blob/73ffa4b7c63bddd5a39b9ecdf35eb3be667f6d21/src/middleware/subscribeWithSelector.ts#L30

Yes, that form is supported.

> and works the same as this one:
> 
> ```ts
> store.subscribe(
>    (store) => store,
>    (current, previous) => console.log(current),
>  );
> ```
> 
> But let's work with the second option — selecting the entire store.
> 
> I'm not sure why selecting the entire store triggers an event when a non-changing update was made, while selecting a property doesn't, that is:
> 
> ```ts
> store.subscribe(
>    (store) => store.a,
>    (current, previous) => console.log(current),
>  );
> 
> // doesn't trigger
> store.setState({a: store.getState().a})
> ```

Because you are passing the new state object.
I think this doesn't trigger in either case:
```js
store.setState(store.getState())
```

> Is this an expected behaviour?

So, yes, it's because the state object is different (not because of the function call form.)

**Reply by ivancuric** [reply]
I think I get it.

You're right, this doesn't trigger it:
```ts
store.setState(store.getState())
```
while the following does - it's a new object:
```ts
store.setState({ ...store.getState() });
```

So to have it not trigger, would I have to use a shallow comparison function when subscribing?

**Reply by dai-shi** [reply, marked as answer]
> So to have it not trigger, would I have to use a shallow comparison function when subscribing?

It depends on your goal.
I think triggering a callback (that does nothing based on a condition) can be more lightweight than a shallow comparison.
(but, in general, we need to "measure" to discuss performance. all I'm saying is hypothetical or intuitive.)
